test/hotspot/jtreg/runtime/Monitor/TestRecursiveLocking.java

```diff
@@ -0,0 +1,529 @@
+ *     TestRecursiveLocking 5 1
+ *
+ * @run main/othervm -Xbootclasspath/a:.
+ *     -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI
+ *     -XX:-EliminateNestedLocks
+ *     -XX:LockingMode=2
+ *     -ms256m -mx256m
+ *     TestRecursiveLocking 5 1
+ */
+
+/*
+ * @test id=C2_alternate_AB
+ * @requires vm.flagless
+ * @requires vm.compiler2.enabled
+ * @summary Tests recursive locking in C2 in alternate A and B mode.
+ * @library /testlibrary /test/lib
+ * @build jdk.test.whitebox.WhiteBox
+ *
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller jdk.test.whitebox.WhiteBox
+ * @run main/othervm -Xbootclasspath/a:.
+ *     -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI
+ *     -XX:LockingMode=0
+ *     -XX:-EliminateNestedLocks
+ *     -ms256m -mx256m
+ *     TestRecursiveLocking 5 2
+ *
+ * @run main/othervm -Xbootclasspath/a:.
+ *     -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI
+ *     -XX:LockingMode=1
+ *     -XX:-EliminateNestedLocks
+ *     -ms256m -mx256m
+ *     TestRecursiveLocking 5 2
+ *
+ * @run main/othervm -Xbootclasspath/a:.
+ *     -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI
+ *     -XX:LockingMode=2
+ *     -XX:-EliminateNestedLocks
+ *     -ms256m -mx256m
+ *     TestRecursiveLocking 5 2
+ */
+
+import jdk.test.lib.Asserts;
+import jdk.test.whitebox.WhiteBox;
+import jtreg.SkippedException;
+
+public class TestRecursiveLocking {
+    static final WhiteBox WB = WhiteBox.getWhiteBox();
+    static final int LockingMode = WB.getIntVMFlag("LockingMode").intValue();
+    static final int LockStackCapacity = WB.getLockStackCapacity();
+    static final int LM_MONITOR = 0;
+    static final int LM_LEGACY = 1;
+    static final int LM_LIGHTWEIGHT = 2;
+    static final SyncThread syncThread = new SyncThread();
+
+    // This SynchronizedObject class and the OUTER followed by INNER testing
+    // model is adapted from runtime/lockStack/TestLockStackCapacity.java.
+    static class SynchronizedObject {
+        private int counter;
+
+        synchronized void runInner(int depth, SynchronizedObject outer) {
+            counter++;
+
+            if (LockingMode != LM_LEGACY) {
+                outer.assertInflated();
+            }
+
+            if (LockingMode != LM_MONITOR) {
+                assertNotInflated();
+            }
+            if (depth == 1) {
+                return;
+            } else {
+                runInner(depth - 1, outer);
+            }
+            if (LockingMode != LM_MONITOR) {
+                assertNotInflated();
+            }
+        }
+
+        synchronized void runOuter(int depth, SynchronizedObject inner) {
+            counter++;
+            if (LockingMode != LM_MONITOR) {
+                assertNotInflated();
+            }
+            if (depth == 1) {
+                inner.runInner(LockStackCapacity, this);
+            } else {
+                runOuter(depth - 1, inner);
+            }
+            if (LockingMode != LM_LEGACY) {
+                assertInflated();
```

Comment:
Legacy isn't inflated because it's recursive?  Can you add a comment why? It would help if something changes and the test starts failing to help whoever breaks this.

Reply:
I'm adding comments for most of the assertions.